### ISSUE TITLE
undefined symbol thrown

### ISSUE THREAD

**Issue by unorthodoxgeek** [mariano/node-db-mysql#16]
node.js:183
        throw e; // process.nextTick error, or 'error' event on first tick
        ^
Error: /srv/sir2/app/node/node_modules/db-mysql/build/default/mysql_bindings.node: undefined symbol: ev_rt_now
    at Object..node (module.js:435:11)
    at Module.load (module.js:339:31)
    at Function._load (module.js:298:12)
    at require (module.js:367:19)
    at Object.<anonymous> (/srv/sir2/app/node/node_modules/db-mysql/db-mysql.js:12:15)
    at Module._compile (module.js:423:26)
    at Object..js (module.js:429:10)
    at Module.load (module.js:339:31)
    at Function._load (module.js:298:12)
    at require (module.js:367:19)

I have libmysqlclient-dev installed and set the MYSQL_CONFIG variable correctly before installing, but still no luck. what am I doing wrong?

**Comment by unorthodoxgeek** [comment]
the output above is the error message I get when trying to run a node.js app (via node app.js)

installing with npm works, sort-of, but there is an error (node path not found), this is the output when installing:

Checking for program g++ or c++          : /usr/bin/g++ 
Checking for program cpp                 : /usr/bin/cpp 
Checking for program ar                  : /usr/bin/ar 
Checking for program ranlib              : /usr/bin/ranlib 
Checking for g++                         : ok  
Checking for node path                   : not found 
Checking for node prefix                 : ok /usr/local 
Checking for program mysql_config        : /usr/bin/mysql_config 
Checking for library mysqlclient_r       : yes 
'configure' finished successfully (0.947s)
Waf: Entering directory `/srv/sir2/app/node/node_modules/db-mysql/build'
[ 1/11] cxx: lib/node-db/binding.cc -> build/default/lib/node-db/binding_1.o
[ 2/11] cxx: lib/node-db/connection.cc -> build/default/lib/node-db/connection_1.o                                                                   
[ 3/11] cxx: lib/node-db/exception.cc -> build/default/lib/node-db/exception_1.o                                                                     
[ 4/11] cxx: lib/node-db/query.cc -> build/default/lib/node-db/query_1.o                                                                             
[ 5/11] cxx: lib/node-db/result.cc -> build/default/lib/node-db/result_1.o                                                                           
[ 6/11] cxx: src/connection.cc -> build/default/src/connection_1.o                                                                                   
[ 7/11] cxx: src/mysql.cc -> build/default/src/mysql_1.o                                                                                             
[ 8/11] cxx: src/query.cc -> build/default/src/query_1.o                                                                                             
/usr/local/include/node/ev/ev.h:565:1: warning: ‘ev_tstamp ev_now()’ defined but not used                                                            
/usr/local/include/node/ev/ev.h:574:1: warning: ‘int ev_is_default_loop()’ defined but not used
/usr/local/include/node/ev/ev.h:810:20: warning: ‘void ev_loop(int)’ defined but not used
/usr/local/include/node/ev/ev.h:811:20: warning: ‘void ev_unloop(int)’ defined but not used
/usr/local/include/node/ev/ev.h:812:20: warning: ‘void ev_default_destroy()’ defined but not used
/usr/local/include/node/ev/ev.h:813:20: warning: ‘void ev_default_fork()’ defined but not used
/usr/local/include/node/ev/ev.h:815:30: warning: ‘unsigned int ev_loop_count()’ defined but not used
/usr/local/include/node/ev/ev.h:816:30: warning: ‘unsigned int ev_loop_depth()’ defined but not used
/usr/local/include/node/ev/ev.h:817:30: warning: ‘void ev_loop_verify()’ defined but not used
[ 9/11] cxx: src/result.cc -> build/default/src/result_1.o
/usr/local/include/node/ev/ev.h:565:1: warning: ‘ev_tstamp ev_now()’ defined but not used                                                            
/usr/local/include/node/ev/ev.h:574:1: warning: ‘int ev_is_default_loop()’ defined but not used
/usr/local/include/node/ev/ev.h:810:20: warning: ‘void ev_loop(int)’ defined but not used
/usr/local/include/node/ev/ev.h:811:20: warning: ‘void ev_unloop(int)’ defined but not used
/usr/local/include/node/ev/ev.h:812:20: warning: ‘void ev_default_destroy()’ defined but not used
/usr/local/include/node/ev/ev.h:813:20: warning: ‘void ev_default_fork()’ defined but not used
/usr/local/include/node/ev/ev.h:815:30: warning: ‘unsigned int ev_loop_count()’ defined but not used
/usr/local/include/node/ev/ev.h:816:30: warning: ‘unsigned int ev_loop_depth()’ defined but not used
/usr/local/include/node/ev/ev.h:817:30: warning: ‘void ev_loop_verify()’ defined but not used
/usr/local/include/node/ev/ev.h:565:1: warning: ‘ev_tstamp ev_now()’ defined but not used
/usr/local/include/node/ev/ev.h:574:1: warning: ‘int ev_is_default_loop()’ defined but not used
/usr/local/include/node/ev/ev.h:810:20: warning: ‘void ev_loop(int)’ defined but not used
/usr/local/include/node/ev/ev.h:811:20: warning: ‘void ev_unloop(int)’ defined but not used
/usr/local/include/node/ev/ev.h:812:20: warning: ‘void ev_default_destroy()’ defined but not used
/usr/local/include/node/ev/ev.h:813:20: warning: ‘void ev_default_fork()’ defined but not used
/usr/local/include/node/ev/ev.h:815:30: warning: ‘unsigned int ev_loop_count()’ defined but not used
/usr/local/include/node/ev/ev.h:816:30: warning: ‘unsigned int ev_loop_depth()’ defined but not used
/usr/local/include/node/ev/ev.h:817:30: warning: ‘void ev_loop_verify()’ defined but not used
[10/11] cxx: src/mysql_bindings.cc -> build/default/src/mysql_bindings_1.o
/usr/local/include/node/ev/ev.h:565:1: warning: ‘ev_tstamp ev_now()’ defined but not used                                                            
/usr/local/include/node/ev/ev.h:574:1: warning: ‘int ev_is_default_loop()’ defined but not used
/usr/local/include/node/ev/ev.h:810:20: warning: ‘void ev_loop(int)’ defined but not used
/usr/local/include/node/ev/ev.h:811:20: warning: ‘void ev_unloop(int)’ defined but not used
/usr/local/include/node/ev/ev.h:812:20: warning: ‘void ev_default_destroy()’ defined but not used
/usr/local/include/node/ev/ev.h:813:20: warning: ‘void ev_default_fork()’ defined but not used
/usr/local/include/node/ev/ev.h:815:30: warning: ‘unsigned int ev_loop_count()’ defined but not used
/usr/local/include/node/ev/ev.h:816:30: warning: ‘unsigned int ev_loop_depth()’ defined but not used
/usr/local/include/node/ev/ev.h:817:30: warning: ‘void ev_loop_verify()’ defined but not used
/usr/local/include/node/ev/ev.h:565:1: warning: ‘ev_tstamp ev_now()’ defined but not used
/usr/local/include/node/ev/ev.h:574:1: warning: ‘int ev_is_default_loop()’ defined but not used
/usr/local/include/node/ev/ev.h:810:20: warning: ‘void ev_loop(int)’ defined but not used
/usr/local/include/node/ev/ev.h:811:20: warning: ‘void ev_unloop(int)’ defined but not used
/usr/local/include/node/ev/ev.h:812:20: warning: ‘void ev_default_destroy()’ defined but not used
/usr/local/include/node/ev/ev.h:813:20: warning: ‘void ev_default_fork()’ defined but not used
/usr/local/include/node/ev/ev.h:815:30: warning: ‘unsigned int ev_loop_count()’ defined but not used
/usr/local/include/node/ev/ev.h:816:30: warning: ‘unsigned int ev_loop_depth()’ defined but not used
/usr/local/include/node/ev/ev.h:817:30: warning: ‘void ev_loop_verify()’ defined but not used
[11/11] cxx_link: build/default/lib/node-db/binding_1.o build/default/lib/node-db/connection_1.o build/default/lib/node-db/exception_1.o build/default/lib/node-db/query_1.o build/default/lib/node-db/result_1.o build/default/src/connection_1.o build/default/src/mysql_1.o build/default/src/query_1.o build/default/src/result_1.o build/default/src/mysql_bindings_1.o -> build/default/mysql_bindings.node                                              
Waf: Leaving directory`/srv/sir2/app/node/node_modules/db-mysql/build'  
'build' finished successfully (2.765s)

**Comment by mariano** [comment]
That looks like it's failing all over the place, particularly with node's headers.

What node version are you using?

**Comment by HendrikN** [comment]
Im having the same issue; fresh new Node.js install (v0.5.0-pre); when starting the app:

```
node.js:183
        throw e; // process.nextTick error, or 'error' event on first tick
        ^
Error: /var/node/app/node_modules/db-mysql/build/default/mysql_bindings.node: undefined symbol: ev_rt_now
    at Object..node (module.js:435:11)
    at Module.load (module.js:339:31)
    at Function._load (module.js:298:12)
    at require (module.js:367:19)
    at Object.<anonymous> (/var/node/app/node_modules/db-mysql/db-mysql.js:12:15)
    at Module._compile (module.js:423:26)
    at Object..js (module.js:429:10)
    at Module.load (module.js:339:31)
    at Function._load (module.js:298:12)
    at require (module.js:367:19)
```

When NPM installed db-mysql:

```
> db-mysql@0.6.5 preuninstall /var/node/app/node_modules/db-mysql
> rm -rf build/*


> db-mysql@0.6.5 install /var/node/app/node_modules/db-mysql
> node-waf configure build

Checking for program g++ or c++          : /usr/bin/g++
Checking for program cpp                 : /usr/bin/cpp
Checking for program ar                  : /usr/bin/ar
Checking for program ranlib              : /usr/bin/ranlib
Checking for g++                         : ok
Checking for node path                   : ok /home/hendrik/local/node/
Checking for node prefix                 : ok /home/hendrik/local/node
Checking for program mysql_config        : /usr/bin/mysql_config
Checking for library mysqlclient_r       : yes
'configure' finished successfully (1.667s)
Waf: Entering directory `/var/node/app/node_modules/db-mysql/build'
[ 1/11] cxx: lib/node-db/binding.cc -> build/default/lib/node-db/binding_1.o
/home/hendrik/local/node/include/node/ev/ev.h:565:1: warning: âev_tstamp ev_now()â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:574:1: warning: âint ev_is_default_loop()â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:810:20: warning: âvoid ev_loop(int)â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:811:20: warning: âvoid ev_unloop(int)â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:812:20: warning: âvoid ev_default_destroy()â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:813:20: warning: âvoid ev_default_fork()â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:815:30: warning: âunsigned int ev_loop_count()â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:816:30: warning: âunsigned int ev_loop_depth()â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:817:30: warning: âvoid ev_loop_verify()â defined but not used
[ 2/11] cxx: lib/node-db/connection.cc -> build/default/lib/node-db/connection_1.o
[ 3/11] cxx: lib/node-db/exception.cc -> build/default/lib/node-db/exception_1.o
[ 4/11] cxx: lib/node-db/query.cc -> build/default/lib/node-db/query_1.o
/home/hendrik/local/node/include/node/ev/ev.h:565:1: warning: âev_tstamp ev_now()â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:574:1: warning: âint ev_is_default_loop()â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:810:20: warning: âvoid ev_loop(int)â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:811:20: warning: âvoid ev_unloop(int)â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:812:20: warning: âvoid ev_default_destroy()â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:813:20: warning: âvoid ev_default_fork()â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:815:30: warning: âunsigned int ev_loop_count()â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:816:30: warning: âunsigned int ev_loop_depth()â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:817:30: warning: âvoid ev_loop_verify()â defined but not used
[ 5/11] cxx: lib/node-db/result.cc -> build/default/lib/node-db/result_1.o
[ 6/11] cxx: src/connection.cc -> build/default/src/connection_1.o
[ 7/11] cxx: src/mysql.cc -> build/default/src/mysql_1.o
/home/hendrik/local/node/include/node/ev/ev.h:565:1: warning: âev_tstamp ev_now()â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:574:1: warning: âint ev_is_default_loop()â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:810:20: warning: âvoid ev_loop(int)â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:811:20: warning: âvoid ev_unloop(int)â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:812:20: warning: âvoid ev_default_destroy()â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:813:20: warning: âvoid ev_default_fork()â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:815:30: warning: âunsigned int ev_loop_count()â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:816:30: warning: âunsigned int ev_loop_depth()â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:817:30: warning: âvoid ev_loop_verify()â defined but not used
[ 8/11] cxx: src/query.cc -> build/default/src/query_1.o
/home/hendrik/local/node/include/node/ev/ev.h:565:1: warning: âev_tstamp ev_now()â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:574:1: warning: âint ev_is_default_loop()â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:810:20: warning: âvoid ev_loop(int)â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:811:20: warning: âvoid ev_unloop(int)â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:812:20: warning: âvoid ev_default_destroy()â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:813:20: warning: âvoid ev_default_fork()â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:815:30: warning: âunsigned int ev_loop_count()â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:816:30: warning: âunsigned int ev_loop_depth()â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:817:30: warning: âvoid ev_loop_verify()â defined but not used
[ 9/11] cxx: src/result.cc -> build/default/src/result_1.o
[10/11] cxx: src/mysql_bindings.cc -> build/default/src/mysql_bindings_1.o
/home/hendrik/local/node/include/node/ev/ev.h:565:1: warning: âev_tstamp ev_now()â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:574:1: warning: âint ev_is_default_loop()â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:810:20: warning: âvoid ev_loop(int)â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:811:20: warning: âvoid ev_unloop(int)â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:812:20: warning: âvoid ev_default_destroy()â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:813:20: warning: âvoid ev_default_fork()â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:815:30: warning: âunsigned int ev_loop_count()â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:816:30: warning: âunsigned int ev_loop_depth()â defined but not used
/home/hendrik/local/node/include/node/ev/ev.h:817:30: warning: âvoid ev_loop_verify()â defined but not used
[11/11] cxx_link: build/default/lib/node-db/binding_1.o build/default/lib/node-db/connection_1.o build/default/lib/node-db/exception_1.o build/default/lib/node-db/query_1.o build/default/lib/node-db/result_1.o build/default/src/connection_1.o build/default/src/mysql_1.o build/default/src/query_1.o build/default/src/result_1.o build/default/src/mysql_bindings_1.o -> build/default/mysql_bindings.node
Waf: Leaving directory `/var/node/app/node_modules/db-mysql/build'
'build' finished successfully (5.192s)
```

**Comment by mariano** [comment]
node.js 0.5.x is not ready for prime time, and I haven't tested node-db on it yet.

Can you switch to latest 0.4.x release and let me know if it works?

**Comment by HendrikN** [comment]
Hmm, downgraded to v0.4.9-pre (you might what to put that in the readme?).

No strange warnings when I NPM the package, however; now I'm getting a Segmantation Fault when i connect to a database. That must be the native MySQL-binaries, right?

**Comment by mariano** [comment]
It could be.

Do me a favor (you must have gdb in your system, and you probably do). If your the file you are using to connect is called test.js, then do the following:

``` bash
$ gdb --args node test.js
```

After which you'll see a command prompt (normally gdb>) Enter these commands (obviously the gdb> part is the prompt, so you shouldn't type it)

```
gdb> run
gdb> bt
```

So after the run command your script should execute, and issue the segmentation fault. Then the bt will print out a backtrace. Could you paste that for me?

**Comment by HendrikN** [comment]
```
GNU gdb (Ubuntu/Linaro 7.2-1ubuntu11) 7.2
Copyright (C) 2010 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>...
Reading symbols from /home/hendrik/local/node/bin/node...done.
(gdb) run
Starting program: /home/hendrik/local/node/bin/node app.js
[Thread debugging using libthread_db enabled]
[New Thread 0x7fffd19f4700 (LWP 8058)]
[Thread 0x7fffd19f4700 (LWP 8058) exited]
[New Thread 0x7fffd0fe6700 (LWP 8059)]

Program received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffd0fe6700 (LWP 8059)]
0x00007fffd1a59aa2 in my_stat () from /usr/lib/libmysqlclient_r.so.16
(gdb) bt
#0  0x00007fffd1a59aa2 in my_stat () from /usr/lib/libmysqlclient_r.so.16
#1  0x00007fffd1a5121a in ?? () from /usr/lib/libmysqlclient_r.so.16
#2  0x00007ffff68392c3 in pthread_once () from /lib/x86_64-linux-gnu/libpthread.so.0
#3  0x00007fffd1a516b4 in get_charset_by_csname () from /usr/lib/libmysqlclient_r.so.16
#4  0x00007fffd1a7255b in mysql_init_character_set () from /usr/lib/libmysqlclient_r.so.16
#5  0x00007fffd1a72d7b in mysql_real_connect () from /usr/lib/libmysqlclient_r.so.16
#6  0x00007fffd1e2e074 in node_db_mysql::Connection::open (this=0xc66670) at ../src/connection.cc:110
#7  0x00007fffd1e18c63 in node_db::Binding::connect (request=0xc72dd0) at ../lib/node-db/binding.cc:104
#8  0x00007fffd1e19354 in node_db::Binding::eioConnect (eioRequest=0xc5ade0) at ../lib/node-db/binding.cc:148
#9  0x0000000000538b78 in eio_execute (thr_arg=0xc5ae90) at ../deps/libeio/eio.c:1826
#10 etp_proc (thr_arg=0xc5ae90) at ../deps/libeio/eio.c:1635
#11 0x00007ffff6832d8c in start_thread () from /lib/x86_64-linux-gnu/libpthread.so.0
#12 0x00007ffff657e04d in clone () from /lib/x86_64-linux-gnu/libc.so.6
#13 0x0000000000000000 in ?? ()
(gdb)
```

There you go :)

**Comment by mariano** [comment]
Wow, wicked. Could you paste the JS code?

**Comment by HendrikN** [comment]
```
var mysql = require('db-mysql');

new mysql.Database({
    hostname:   'thehostname',
    user:       'theuser',
    password:   'mysuperverysecretpass',
    database:   'databasename'
}).connect(function(error) {
    if (error) {
        return console.log("CONNECTION ERROR: " + error);
    }
});
```

Stripped the app.js down to just that, still the segfault...

**Comment by mariano** [comment]
What Ubuntu is this on (version, and architecture)? Version of mysql?

One of my virtual machines is ubuntu 11.04 64 bits, with mysql version 5.1.54 as it comes from ubuntu's repos, and the following JS (which is almost exactly like yours) works great:

``` javascript
var mysql=require('db-mysql');

new mysql.Database({
    hostname: 'localhost',
    user: 'root',
    password: 'password',
    database: 'node'
}).connect(function(error) {
    if (error) {
        return console.log('CONNECT ERROR: ' + error);
    }
    this.query().select('*').from('users').execute(function(error, rows) {
        if (error) {
            return console.log('QUERY ERROR: ' + error);
        }
        console.log(rows[0]);
        console.log(rows.length + ' ROWS');
    });
});
```

**Comment by HendrikN** [comment]
Ubuntu 11.04 (x64). 

MySQL-version (on the VM that is running node) is ```
mysql  Ver 14.14 Distrib 5.1.54, for debian-linux-gnu (x86_64) using readline 6.2

```

I am connecting to an external host, but that shouldn't be a problem. (and yes, i can ping to that server from the node-VM). Thanks for your effort :)
```

**Comment by mariano** [comment]
Yeah remote server has nothing to do with it, since i just changed the test.js i pasted above to connect to a remote server, and it works

**Comment by mariano** [comment]
And connecting via the mysql client binary from the node-VM to the mysql server works ok?

**Comment by HendrikN** [comment]
Okay...

Connecting from the VM to the external DB with mysql works also:
```mysql --host=thehostname --user=theuser --password=topsecret dbname

> show tables;
> 
> ```
> ```

So.. eh...

**Comment by mariano** [comment]
Humor me. Set the async option to false in your connection options, since it seems to be a threading issue (within libmysql) that for some reason only happens to you. So your Database instantiation would be left as:

``` javascript
new mysql.Database({
    async: false,
    hostname:   'thehostname',
    user:       'theuser',
    password:   'mysuperverysecretpass',
    database:   'databasename'
})
```

**Comment by mariano** [comment]
Wow, I think I am able to replicate this on Ubuntu 32 bits...

**Comment by HendrikN** [comment]
Sorry, still the segfault. Setting up a new VM as we speak.

**Comment by mariano** [comment]
Yes, in ubuntu 32 bits I can replicate every time... Damn mysql... Ok I'll work on a fix tomorrow and see what's up.

If on your other VM it works, let me know.

**Comment by HendrikN** [comment]
New VM Installed, same setup (Ubuntu Server 11.04 x64). Clean install, just Node + libmysqlclient-dev + db-mysql... And the Segmentation Fault is still there.

```
var mysql = require('db-mysql');

var db = new mysql.Database({
    hostname:   'host',
    user:       'user',
    password:   'passwd',
    async: false,
    database:   'dbname'
});

db.connect();
```

**Comment by mariano** [comment]
Fixed in 0.6.6. [See release announcement](http://nodejsdb.org/2011/06/db-mysql-0-6-6-released/) Thanks for helping me debug this!

**Comment by HendrikN** [comment]
Thanks for fixing! Works like a charm :)
